Pass exception object with event when throwing exception

emil-furniss opened the pull request:
For any non-`ok` response, pass the entire error object to the UI to enable it to respond accordingly.

mdgbayly commented on store/siren-action-behavior.html:
```diff
@@ -100,9 +100,15 @@
+						return resp.json().then(function(data) {
+							var obj = JSON.parse(data);
+							var errMsg = resp.statusText + ' response executing ' + opts.method + ' on ' + href + '.';
+							throw { object: obj, message: errMsg };
```
@awikkerink @alogan519 

Just as background, the reason for these changes is that we want to start returning more detailed error responses from some of our APIs that will be rendered as messages in the client.

We're using a SirenHypermediaResultProcessor in the LMS to generate siren error messages that aligned to the HTTP Problem Detail spec rfc.

An example looks like this. We plan on localizing the `detail` property.

```
{
    "class": [
        "error"
    ],
    "properties": {
        "status": "BadRequest",
        "code": 400,
        "type": "https://rubrics.api.brightspace.com/rels/errors/out-of-range-points-error",
        "title": "Out of Range Points Error",
        "detail": "Point value must be greater than or equal to 0 and less than or equal to 100,000"
    }
}
```

emil-furniss commented:
I've moved the `catch` block out again to catch and simplified the error being passed to the event. 
Responses that don't have a json object don't cause `resp.json()` to be rejected, but rather `resp.json` is simply not a function in those cases which is handled with an `if (resp.json)` query.
Both cases (with json and without) now end up in the same `catch` block, pass the entire error object to the event (to be handled by the UI), and then throw the same formatted error string as before.

awikkerink commented on store/siren-action-behavior.html:
```diff
@@ -101,8 +101,14 @@
 						var errMsg = resp.statusText + ' response executing ' + opts.method + ' on ' + href + '.';
-						var error = new Error(errMsg);
-						throw error;
+						if (resp.json) {
```
The `json` method should always be available. I'm not sure how to check if the type is json parseable.

mdgbayly replied:
Yeh, I'm surprised that the `json` method isn't always avaialble, but Emil seemed to be saying that in his testing it appeared it wasn't if the body didn't have a json representation. @emil-furniss can you double check that because according to the MDN docs it should always be there. It isn't clear from those docs thought what happens if the body is not parseable as JSON.

Also it seems like the result of calling `json()` should be an object, so it would seem strange that we would have to call `JSON.parse(data)` as it should already be parsed into an object.

emil-furniss replied:
It wasn't clear to me from the docs if the `json()` method is always available - I'll remove that check if you say it is. 
And the result of calling `json()` here is a string, so we need `JSON.parse`. If the string is not parsable, that method will fail with a `SyntaxError`.

mdgbayly commented on store/siren-action-behavior.html:
```diff
@@ -123,9 +129,9 @@
-					self.fire('d2l-siren-entity-save-error', { error: new Error(reason.message) });
-					throw reason;
-				});
+					self.fire('d2l-siren-entity-save-error', reason);
```
Just an FYI - we have code in d2l-rubrics that looks for the `error` property of the event and I think just includes it as a hidden field when it shows the event. 

Also I think we should just throw the original reason as before, otherwise there is a possibility we slice off other information that may have been included in the original error reason.

For similar reasons, I don't think we need to rewrap the error object we add to the event.
We could just do 
`self.fire('d2l-siren-entity-save-error', { error: reason });`

I don't particularly care if we just throw the original parsed JSON object or we wrap it in another object with the JSON object and the formatted errMsg.

emil-furniss commented:
Now catching failed JSON parses and including as much information as possible in the error object, and throwing the full error object at the end.
